pkg/scheduler/factory/framework.go

```diff
@@ -0,0 +1,95 @@
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/scheduler/apis/config"
+)
+
+// FrameworkConfigProducer produces a framework's configuration.
+type FrameworkConfigProducer func() (config.Plugins, []config.PluginConfig)
+
+// FrameworkConfigProducerRegistry tracks mappings from predicates/priorities to framework config producers.
+type FrameworkConfigProducerRegistry struct {
+	// maps that associate predicates/priorities with framework plugin configurations.
+	predicateToFrameworkConfigProducer map[string]FrameworkConfigProducer
+	priorityToFrameworkConfigProducer  map[string]FrameworkConfigProducer
+}
+
+// NewFrameworkConfigProducerRegistry creates a new producer registry.
+func NewFrameworkConfigProducerRegistry() *FrameworkConfigProducerRegistry {
+	return &FrameworkConfigProducerRegistry{
+		predicateToFrameworkConfigProducer: make(map[string]FrameworkConfigProducer),
+		priorityToFrameworkConfigProducer:  make(map[string]FrameworkConfigProducer),
+	}
+}
+
+func registerProducer(name string, producer FrameworkConfigProducer, producersMap map[string]FrameworkConfigProducer) {
+	validateAlgorithmNameOrDie(name)
+	producersMap[name] = producer
+}
+
+// RegisterPredicate registers a config producer for a predicate.
+func (f *FrameworkConfigProducerRegistry) RegisterPredicate(name string, producer FrameworkConfigProducer) {
+	registerProducer(name, producer, f.predicateToFrameworkConfigProducer)
+}
+
+// RegisterPriority registers a framework config producer for a priority.
+func (f *FrameworkConfigProducerRegistry) RegisterPriority(name string, producer FrameworkConfigProducer) {
+	registerProducer(name, producer, f.priorityToFrameworkConfigProducer)
+}
+
+// IsPredicateRegistered returns true if the key exist in the predicates registry
+func (f *FrameworkConfigProducerRegistry) IsPredicateRegistered(name string) bool {
+	_, exist := f.predicateToFrameworkConfigProducer[name]
+	return exist
+}
+
+// IsPriorityRegistered returns true if the key exist in the priority map
+func (f *FrameworkConfigProducerRegistry) IsPriorityRegistered(name string) bool {
+	_, exist := f.priorityToFrameworkConfigProducer[name]
+	return exist
+}
+
+func produceFrameworkConfig(keys []string, producersMap map[string]FrameworkConfigProducer) (*config.Plugins, []config.PluginConfig, error) {
+	var plugins config.Plugins
+	var pluginConfig []config.PluginConfig
+	for _, k := range keys {
+		producer, exist := producersMap[k]
+		if !exist {
+			return nil, nil, fmt.Errorf("finding key %q", k)
+		}
+		p, pc := producer()
+		plugins.Append(&p)
+		pluginConfig = append(pluginConfig, pc...)
+	}
+	return &plugins, pluginConfig, nil
+}
+
+// ProducePredicateFrameworkConfig produces framework configuration for the list of provided
+// predicate keys. Corresponding producers should have been registered before calling this function.
+func (f *FrameworkConfigProducerRegistry) ProducePredicateFrameworkConfig(predicateKeys []string) (*config.Plugins, []config.PluginConfig, error) {
```

Comment:
What's the benefit of having separated exported functions instead of a single `ProduceFrameworkConfig`?

Reply:
I was trying to encapsulate access to the maps so that the caller deals with the struct methods only to hide the implementation detail of the registry (this is all C++ style). In any case, I removed this because the factory will want to pass customized arguments for each producer, and so there is point in having this function anymore.